samples/rust/rust_platdev.rs

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Rust platform device sample
+
+#![no_std]
+#![feature(allocator_api, global_asm)]
+
+use alloc::{boxed::Box, sync::Arc};
+use core::pin::Pin;
+use kernel::prelude::*;
+use kernel::{
+    cstr,
+    file::File,
+    file_operations::{FileOpener, FileOperations},
+    io_buffer::IoBufferWriter,
+    miscdev,
+    platform_driver::{self, PlatformDevice, PlatformDriver},
+    regmap::{Regmap, RegmapConfig},
+};
+
+module! {
+    type: RustPlatdev,
+    name: b"rust_platdev",
+    author: b"Rust for Linux Contributors",
+    description: b"Rust platform device sample",
+    license: b"GPL v2",
+    params: {
+    },
+}
+
+// The Shared State is simply a Regmap, which is Send + Sync.
+struct SharedState {
+    regmap: Regmap,
+}
+
+impl SharedState {
+    fn try_new(regmap: Regmap) -> KernelResult<Arc<Self>> {
+        Ok(Arc::try_new(SharedState { regmap })?)
+    }
+}
+
+struct RngDevice {
+    state: Arc<SharedState>,
```

Comment:
I thought about this some more. Many (if not most) drivers use `devm_` lifetime management: kernel objects created using `devm_xxx(struct device *dev, ...)` all have lifetimes that are tied to the lifetime of the `struct device`. Right before the device is released/deallocated by the kernel, all associated objects are automatically cleaned up (in reverse order of creation, and by calling their release functions).

Here is how the `struct regmap`'s lifetime is controlled in our platform driver:
```c
int probe(struct platform_device *pdev)
{
    struct device *dev = &pdev->dev; /* platform device "is-a" device */
    void __iomem *iomem = devm_ioremap(dev, ...);
    struct regmap *regmap = devm_regmap_init_mmio(dev, iomem);
}

int remove(struct platform_device *pdev)
{
    /* no need to explicitly clean up regmap or iomem, or care about the order:
     * the kernel takes care of that right before the platform device is cleaned up
     */
}
```
And now the driver writer no longer has clean up the `struct regmap`, or care about its lifetime: it'll get cleaned up automatically.

So how can we leverage Rust to verify this at build time? Do we confer a lifetime on the `struct device`'s Rust abstraction, which corresponds to the actual kernel device lifetime? And do we transfer that lifetime onto Rust abstractions of kernel objects which are tied to the device lifetime?

Alternatively, we can forego the `devm_` kernel mechanism completely (it's optional), and manually manage our own kernel resources in Rust. Then we still need to watch out for linked lifetimes: the `struct regmap` depends on the `iomem`, for example. And some kernel objects may still depend on a `struct device`, even when not `devm_` managed.

If the Rust abstraction of the `struct regmap` needs to link to its dependencies, things may get messy, because regmaps may be created from various underlying resources: iomem, spi, i2c, clocks... I hope we won't need a `Regmap` type for every regmap variant ?! There are a great many possible combinations.